Add/gated blog domain available for free plan

southp opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to Automattic/martech#1794 

## Proposed Changes

This PR implements the foundational control/treatment experience of the experiment proposed in Automattic/martech#1794. Instead of incorporating ExPlat directly, it uses a feature flag instead so it can be iterated further in production if needed, easier to test, and better focus on the UX and the general code structure to begin with.

It is not written in a throwable way like we often did in the past for A/B tests. Instead, it attempts to introduce a basic structure that enables future experiments on paid domains + the Free plan. For example, it should be intuitive to expand into peP6yB-lq-p2. This is also a test to the idea of "Explore What Enables More Efficient Growth Experiments" I outlined in pdgrnI-2vQ-p2: instead of writing throwable code for short-term speed, we should strive more for a flexible strucutre from the real cases so we accumulate the invested effort in each experiment and will move faster and faster, instead of always going back to square 1 after each.

The treatment experience can be found in the feature list:
<img width="828" alt="image" src="https://github.com/Automattic/wp-calypso/assets/1842898/cbf5c358-7aa6-4fad-8a6e-31bb52161e0e">

and the paid domain + free plan modal:
<img width="1271" alt="image" src="https://github.com/Automattic/wp-calypso/assets/1842898/85cd9983-8559-4efc-a248-f90c97769ff5">

## Testing Instructions
<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

#### Treatment experience

* Enable the treatment experience by either running `ENABLE_FEATURES=domains/blog-domain-free-plan yarn start` to start your local calypso, or appending `?flags=domains/blog-domain-free-plan`
* Go through the main onboarding flow, and pick a .blog domain in the domain step.
* In the plans page, it should say "xxx.blog is a redirect"
* Clicking the Free plan button, the modal should show the new copy of explaining why a paid plan is required for a custom primary domain.
* Make sure both the CTAs work as expected.
* Pick a non-blog paid domain, and it should act just like what we have in production now.

#### Control experience

By not doing the priliminary feature enabling step, you should be able to access the default experience that we have now. Everything should work as normal.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

aneeshd16 commented on client/my-sites/plan-features-2023-grid/components/features.tsx:
```diff
@@ -84,17 +106,7 @@ const PlanFeatures2023GridFeatures: React.FC< {
-												domainName={ domainName as string }
-											/>
-										</Plans2023Tooltip>
+										<FreePlanCustomDomainFeature key={ key } domainName={ domainName as string } />
```
Non-blocking: Just curious why the tooltip was moved inside the `FreePlanCustomDomainFeature` component?

southp replied:
Good point. Initially I just did that thinking it might make it look cleaner. Relooking at it, leaving the tooltip there is actually better since it keeps the DOM structure of different conditional branches the same. I've brought it back :)

aneeshd16 commented on client/my-sites/plans-features-main/components/free-plan-paid-domain-dialog.tsx:
```diff
@@ -119,30 +121,82 @@ const StyledButton = styled( Button )`
 `;
 
-export function FreePlanPaidDomainDialog( {
+function GetSuggestedPlanSection( {
```
Nit: Since this is a component, can we consider renaming this to `SuggestedPlanSection`?

southp replied:
Done ;)

aneeshd16 commented on client/my-sites/plans-features-main/components/free-plan-paid-domain-dialog.tsx:
```diff
@@ -119,30 +121,82 @@ const StyledButton = styled( Button )`
+		<>
+			<DomainName>
+				<div>{ domainName }</div>
+				<FreeDomainText>{ translate( 'Free for one year ' ) }</FreeDomainText>
```
Just noticed that there was an extra space here. (Was present earlier too)
```suggestion
				<FreeDomainText>{ translate( 'Free for one year' ) }</FreeDomainText>
```

southp replied:
Good catch! I've committed the fix :)

aneeshd16 commented on client/my-sites/plans-features-main/components/free-plan-paid-domain-dialog.tsx:
```diff
@@ -153,11 +207,138 @@ export function FreePlanPaidDomainDialog( {
+
+	function handleFreePlanClick() {
+		setIsBusy( true );
+		onFreePlanSelected();
```
If the WPCOM domain name is not passed to `onFreePlanSelected`, this will lead to a mismatch between the domain we show in the modal and the actual address of the site which will be created.

https://github.com/Automattic/wp-calypso/assets/5436027/a7316281-1a7b-4ab3-9b3f-88a7b78f1f78

southp replied:
Nice finding. It is addressed in https://github.com/Automattic/wp-calypso/pull/78825/commits/50f29474e1b573023c6684baf6ca0c2c6d8ace05 by separating `replacePaidDomainWithFreeDomain`

aneeshd16 reviewed (commented):
Code looks good and tests out, except for the case I have highlighted below. 👍 

I also have some feedback regarding the modal itself:
1. I'm wondering if we can simplify the modal copy without using the "primary domain" term. This is an unknown term to new users on our platform and might cause confusion. Although, this could be fine since we are already linking to the support document. Sorry, but I don't have any great suggestions for alternative copy either. :/
2. The `<domain> redirects to <wpcom domain>` text is all in the same style, so it could be tough to read. Do you think we can get a design review on this?
<img width="963" alt="image" src="https://github.com/Automattic/wp-calypso/assets/5436027/69eda11d-1586-4079-82a0-4718bca1ef09">

chriskmnds commented on client/my-sites/plan-features-2023-grid/components/features.tsx:
```diff
@@ -4,6 +4,7 @@ import styled from '@emotion/styled';
 import classNames from 'classnames';
 import { useTranslate } from 'i18n-calypso';
 import { LoadingPlaceHolder } from '../../plans-features-main/components/loading-placeholder';
+import useIsCustomDomainAllowedOnFreePlan from '../../plans-features-main/hooks/use-is-custom-domain-allowed-on-free-plan';
```
We shouldn't be importing this in here. Generate the value from plans-features-main and pass as **optional** prop to plans-features-2023-grid

southp replied:
Extracted by https://github.com/Automattic/wp-calypso/pull/78825/commits/e20c4cd1be566658fea954b235ab43155e072884

chriskmnds commented on client/my-sites/plan-features-2023-grid/components/features.tsx:
```diff
@@ -28,12 +29,34 @@ const FreePlanCustomDomainFeature: React.FC< { domainName: string } > = ( { doma
 
+	const translate = useTranslate();
+	const isCustomDomainAllowedOnFreePlan = useIsCustomDomainAllowedOnFreePlan( domainName );
+	const wpcomFreeDomain = ! isError && wordPressSubdomainSuggestions?.[ 0 ]?.domain_name;
```
I think these suggestions should really be passed into the plans-features-2023-grid components from outside, either as a list of suggestions or a hook to call. not fond of the network dependency that's hidden inside here.

ddc22 replied:
Maybe the domain 'feature' should be injected as its own component and it can then simply be rendered on the grid 🤔 

The domain feature is going to be experimented heavily as suggested by this PR. And this could be the approach we use to decouple the grid from external dependencies like explat when they are required to be used for experimented and other similar requirements.

southp replied:
Good suggestion! I've addressed this in a separate PR so it's better scoped: https://github.com/Automattic/wp-calypso/pull/79678 and I'd be grateful for a review for that. It's not yet the way suggested by @jdc91 yet, but I think it's a step forward of decoupling the hidden network dependency :)

chriskmnds commented on client/my-sites/plans-features-main/hooks/use-is-custom-domain-allowed-on-free-plan.ts:
```diff
@@ -0,0 +1,14 @@
+import { getTld } from 'calypso/lib/domains';
+
+// TODO
+// it's not really a hook now, but it eventually will contain a request fetching assigned variations from ExPlat endpoints
```
Probably remove this comment. In what way is this _not_ a hook?

southp replied:
Good point. Initially I added that comment since it's just a simple function that didn't do anything that I'd expect from a hook, e.g. side effect computation, state management, etc. However, it is treated as a hook so this comment is meaningless. It is now removed by https://github.com/Automattic/wp-calypso/pull/78825/commits/b2fb1f2363b6581c6485959b2307b7fd2441831a .

chriskmnds reviewed (changes requested):
I haven't tested anything. It looks like @aneeshd16 is doing a more thorough review. 

A couple of comments in the code, and [one that is blocking](https://github.com/Automattic/wp-calypso/pull/78825#discussion_r1261077674) for me. 🙌

jdc91 commented:
Is it expected that this experiment will also run on the paid media flow? @southp

southp commented:
Feedback addressing is in progress, but I have to shift gear to other things now. Memo to myself:

- [ ] Extrat the domain suggestion outside of the features component. (side note: the abstraction is leaky when the call site needs to access react-query interface directly)
- [ ] The Free domain suggestion needs to be somehow passing in to the dependency store without wiping out the domainCartItem when clicking "continue with the Free plan".
- [ ] This experiment should only be run on `onboarding`, not the paid media flow. (h/t @jdc91 for the reminder). It might be simpler to do this in the follow-up PR https://github.com/Automattic/wp-calypso/pull/79279 as part of the `isEligible` prop.

southp commented:
@aneeshd16 Re:

> I'm wondering if we can simplify the modal copy without using the "primary domain" term. This is an unknown term to new users on our platform and might cause confusion. Although, this could be fine since we are already linking to the support document. Sorry, but I don't have any great suggestions for alternative copy either. :/

> The <domain> redirects to <wpcom domain> text is all in the same style, so it could be tough to read. Do you think we can get a design review on this?

These are both great points. However, it's an experiment that doesn't have design resource up-front and it's already delayed, l'd suggest to not run a design review round until the treatment wins as long as it's not terribly wrong. Does that sound fine to you?

southp commented:
Thanks for the thorough review, @aneeshd16 @chriskmnds @jdc91 . All the feedbacks have been addressed either here or in the separate PR: https://github.com/Automattic/wp-calypso/pull/79678. Restricting this into `onboarding` flow only will also be implemented in [the follow-up PR](https://github.com/Automattic/wp-calypso/pull/79279).
 
When y'all have some time, I'd be grateful for a re-review :)

aneeshd16 commented:
> These are both great points. However, it's an experiment that doesn't have design resource up-front and it's already delayed, l'd suggest to not run a design review round until the treatment wins as long as it's not terribly wrong. Does that sound fine to you?

Sure, the design issue can be tackled later. But I was wondering if we can have a copy review for the text since the success of the experiment depends on it. Maybe a quick review while it is being code reviewed? I noticed that the experiment will be launched to an EN-only audience, so there won't be a translation delay.

southp commented:
> Sure, the design issue can be tackled later. But I was wondering if we can have a copy review for the text since the success of the experiment depends on it. Maybe a quick review while it is being code reviewed? I noticed that the experiment will be launched to an EN-only audience, so there won't be a translation delay.

Good point, and thanks for being the guardrail when. I'll ask our designers to see if they have some time to pitch in here :)

southp committed:
Extract the dialog code into <DialogPaidPlanIsRequired> in preparation
of introducing a new .blog + Free domain dialog.

Introduce `<DialogPaidDomainAndFreePlan>`

Dedupe the props of the two dialogs.

Extract the shared logic of getting a suggested plan into
<GetSuggestedPlanSection> component

Edit the description of `DialogPaidDomainAndFreePlan`

Add the minimal version of the hool `useIsCustomDomainAllowedOnFreePlan`

Differ the allowed actions and copies based on
`isCustomDomainAllowedOnFreePlan`

Emphasize that the custom domain will be a redirect in the feature list

Customize the copy and CTA of the .blog + Free plan dialog.

GetSuggestedPlanSection -> SuggestedPlanSection

Remove the redundant trailing space.

Remove the comment that is soon to be meaningless.

Move `useIsCustomDomainAllowedOnFreePlan` out to <PlanFeaturesMain/>

Separate `replacePaidDomainWithFreedomain` into `removePaidDomain` and
`setSiteUrlAsFreeDomain`, so the paid domain can be conditionally
removed.

southp commented:
@aneeshd16 @chriskmnds @jdc91 FYI I did a controversial rebase of squashing all the commits into one, since having to fix merge conflicts of 15 commits is a bit too challenging to keep up. :)

southp commented:
@lucasmendes-design Thanks for the feedback. Since @jdc91 's project is based off from here, I'll address it in a separate PR once this is merged.

southp committed:
wpcomFreeDomainSuggestion.entry needs to be checked before passing to
the site url setting function

chriskmnds commented on client/my-sites/plan-features-2023-grid/index.tsx:
```diff
@@ -111,6 +111,7 @@ export type PlanFeatures2023GridProps = {
 	// Value of the `?feature=` query param, so we can highlight a given feature and hide plans without it.
 	selectedFeature?: string;
 	intent?: PlansIntent;
+	isCustomDomainAllowedOnFreePlan?: boolean | null;
```
Can you help me understand the intention of `isCustomDomainAllowedOnFreePlan`?

Offhand - this prop feels unnecessary if it's something that we can induce by the presence of `wpcomFreeDomainSuggestion.entry` alone 🤔 

We really should be trying to contain things behind as fewer props as possible, where it makes sense. Do `isCustomDomainAllowedOnFreePlan` and `wpcomFreeDomainSuggestion` have any value separately? Is there  any effect on the grid when these are _not_ used in conjunction? Perhaps `undefined` or `null` value for entry should be enough to cover the use case?

p.s. not sure if rethinking the name `wpcomFreeDomainSuggestion` to also convey its intended use, perhaps might help? 🤔

ddc22 replied:
Just to provide some context here `isCustomDomainAllowedOnFreePlan` is a flag introduced to activate an experiment. 

https://github.com/Automattic/wp-calypso/blob/1612be517abe1058343afde8c0d244f76905912c/client/my-sites/plans-features-main/hooks/use-is-custom-domain-allowed-on-free-plan.ts#L9

`wpcomFreeDomainSuggestion` will be generated regardless of the experiment running. On that note I think you zeroed in on a bug here. The below condition needs to be fixed.

https://github.com/Automattic/wp-calypso/blob/fae5f7fa82683c59eaa6c2ee1260785ab3acfb07/client/my-sites/plan-features-2023-grid/components/features.tsx#L34-L46

The entire html segment above should not be shown if `entry` is empty 🤔 . Correct me if I am wrong @southp

southp replied:
Good question @chriskmnds and thanks for pitching in @jdc91 . What's more important than the feature flag is that the hook itself will be replaced by ExPlat integration in https://github.com/Automattic/wp-calypso/pull/79279. I wanted to separate the implementation of this treatment experience and the ExPlat integration out so they can be reviewed and be easier to test. I don't want to go through Abacus everytime I'm testing this :) 

Sorry that I didn't explain that very well in the description. Does it make better sense now?

southp replied:
> wpcomFreeDomainSuggestion will be generated regardless of the experiment running. On that note I think you zeroed in on a bug here. The below condition needs to be fixed.

Great catch @jdc91 . It is now fixed by https://github.com/Automattic/wp-calypso/pull/78825/commits/4304d2396ac5c031f6942168b6e0bbeb309aa6d0 .

chriskmnds replied:
Thanks. It doesn't answer my questions though (although you may have confirmed what I was saying). I am specifically asking and interested in properties as sent through to `plan-features-2023-grid`. I'm a little suspicious if we need two of them to serve a single piece of logic. It creates complexity at several levels on things that are coupled to begin with. 

The options I'm seeing are:

1. Rename (potentially) `wpcomFreeDomainSuggestion` to something more tailored to the use case, and use `isCustomDomainAllowedOnFreePlan` to conditionally populate an entry for it prior to reaching plan-features-2023-grid. This would lead to a couple of states that would serve the logic: 
    - `{ loading: false, entry: null }` nothing to show, don't override
    - `{ loading: true, entry: null }` waiting for something to show
    - `{ loading: true|false, entry: foo }` got "foo" suggestion, render it instead of default
2. Send only `isCustomDomainAllowedOnFreePlan` and let the grid call `useSuggestedFreeDomainFromPaidDomain` to retrieve the suggestions. This is similar to how it was done before (with grid making a network request), but it's obviously better abstracted and controlled now from the underlying refactors. `useSuggestedFreeDomainFromPaidDomain` can move under `plan-features-2023-grid` in that case and eventually into one of the grid packages.

or something else I miss.

chriskmnds replied:
> use `isCustomDomainAllowedOnFreePlan` to conditionally populate an entry for it prior to reaching plan-features-2023-grid.

or to just conditionally send through the whole prop `wpcomFreeDomainSuggestion` (populated with entries or not). The rest should fall through as normal. 

`<OnboardingGrid { ...( isCustomDomainAllowedOnFreePlan && { wpcomFreeDomainSuggestion } ) }`

chriskmnds replied:
ohh wait - I'm seeing the changes now and the two are _not_ actually coupled right?

So scrap all that I wrote in that case. My whole premise was on them being coupled inside `plan-features-2023-grid` 😬

southp replied:
Yes, they are not really coupled and are representing different logics. `wpcomFreeDomainSuggestion` is the wpcom free domain derived from a picked paid domain and it is shown in both the plan features grid and the free plan + paid domain modal. It is needed no matter `isCustomDomainAllowedOnFreePlan` is true or not.
`isCustomDomainAllowedOnFreePlan` represents the case that we allow the combination of a paid domain + a free plan; in this PR it's controlled by a feature flag, and it will be controlled by ExPlat API in [the follow-up PR](https://github.com/Automattic/wp-calypso/pull/79279).

In this sense, how do you see the PR now? :) 

(Btw, I really do like how you review the data design carefully. On an super abstract level, our job is to transform data, so clean code can only happen when the data is clean, not the other way around. So data design review is actually a crucial part of a code review, yet people (including me) often miss :) )

chriskmnds replied:
Sounds good. Thanks for explaining. Let's just add a comment/description next to this on what this property (`isCustomDomainAllowedOnFreePlan`) is meant to do (which we'll use for the grid's README later). 👍

southp replied:
Great suggestion. I've added it by https://github.com/Automattic/wp-calypso/pull/78825/commits/ae284c995f1ecfdb0c19b65e0766bee1499795ac :)

southp committed:
Bug fix: wpcomFreeDomainSuggestion.entry shouldn't block the whole
custom domain feature section from rendering.

southp committed:
Add a comment to explain what `isCustomDomainAllowedOnFreePlan` does

ddc22 reviewed (approved):
I have tested this change and it works well.
LGTM 🚢

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/8632413

Thank you @southp for including a screenshot in the description! This is really helpful for our translators.

chandrawisnu commented:
> Related to Automattic/martech#1794
> 
> ## Proposed Changes
> This PR implements the foundational control/treatment experience of the experiment proposed in Automattic/martech#1794. Instead of incorporating ExPlat directly, it uses a feature flag instead so it can be iterated further in production if needed, easier to test, and better focus on the UX and the general code structure to begin with.
> 
> It is not written in a throwable way like we often did in the past for A/B tests. Instead, it attempts to introduce a basic structure that enables future experiments on paid domains + the Free plan. For example, it should be intuitive to expand into peP6yB-lq-p2. This is also a test to the idea of "Explore What Enables More Efficient Growth Experiments" I outlined in pdgrnI-2vQ-p2: instead of writing throwable code for short-term speed, we should strive more for a flexible strucutre from the real cases so we accumulate the invested effort in each experiment and will move faster and faster, instead of always going back to square 1 after each.
> 
> The treatment experience can be found in the feature list: <img alt="image" width="828" src="https://user-images.githubusercontent.com/1842898/251617391-cbf5c358-7aa6-4fad-8a6e-31bb52161e0e.png">
> 
> and the paid domain + free plan modal: <img alt="image" width="1271" src="https://user-images.githubusercontent.com/1842898/251617422-85cd9983-8559-4efc-a248-f90c97769ff5.png">
> 
> ## Testing Instructions
> #### Treatment experience
> * Enable the treatment experience by either running `ENABLE_FEATURES=domains/blog-domain-free-plan yarn start` to start your local calypso, or appending `?flags=domains/blog-domain-free-plan`
> * Go through the main onboarding flow, and pick a .blog domain in the domain step.
> * In the plans page, it should say "xxx.blog is a redirect"
> * Clicking the Free plan button, the modal should show the new copy of explaining why a paid plan is required for a custom primary domain.
> * Make sure both the CTAs work as expected.
> * Pick a non-blog paid domain, and it should act just like what we have in production now.
> 
> #### Control experience
> By not doing the priliminary feature enabling step, you should be able to access the default experience that we have now. Everything should work as normal.
> 
> ## Pre-merge Checklist
> * [ ]  Has the general commit checklist been followed? (PCYsg-hS-p2)
> * [ ]  [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
> * [ ]  Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
> * [ ]  Have you checked for TypeScript, React or other console errors?
> * [ ]  Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
> * [ ]  Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
> * [ ]  For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/8632413).
